runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java

```diff
@@ -58,6 +86,107 @@ public SamzaRunner(SamzaPipelineOptions options) {
     this.options = options;
   }
 
+  private void closeFnServer(GrpcFnServer<?> fnServer) {
+    try (AutoCloseable closer = fnServer) {
+      // do nothing
+    } catch (Exception e) {
+      LOG.error("Failed to close fn api servers. Ignore since this is shutdown process...", e);
+    }
+  }
+
+  private void setUpContextManager(
+      StreamGraph streamGraph, SamzaExecutionContext executionContext) {
+    streamGraph.withContextManager(
+        new ContextManager() {
+          @Override
+          public void init(Config config, TaskContext context) {
+            if (executionContext.getMetricsContainer() == null) {
+              final MetricsRegistryMap metricsRegistry =
+                  (MetricsRegistryMap) context.getSamzaContainerContext().metricsRegistry;
+              executionContext.setMetricsContainer(new SamzaMetricsContainer(metricsRegistry));
+            }
+
+            if (SamzaRunnerOverrideConfigs.isPortableMode(options)) {
+              if (jobBundleFactory == null) {
+                try {
+                  final long waitTimeoutMs =
+                      SamzaRunnerOverrideConfigs.getControlClientWaitTimeoutMs(options);
+                  final InstructionRequestHandler instructionHandler =
+                      controlClientPool
+                          .getSource()
+                          .take(SAMZA_WORKER_ID, Duration.ofMillis(waitTimeoutMs));
+                  final EnvironmentFactory environmentFactory =
+                      environment -> RemoteEnvironment.forHandler(environment, instructionHandler);
+                  // TODO: use JobBundleFactoryBase.WrappedSdkHarnessClient.wrapping
+                  jobBundleFactory =
+                      SingleEnvironmentInstanceJobBundleFactory.create(
+                          environmentFactory, fnDataServer, fnStateServer);
+                } catch (Exception e) {
+                  throw new RuntimeException(
+                      "Running samza in Beam portable mode but failed to create job bundle factory",
+                      e);
+                }
+                executionContext.setJobBundleFactory(jobBundleFactory);
+              }
+            }
+
+            context.setUserContext(executionContext);
+          }
+
+          @Override
+          public void close() {
+            closeFnServer(fnControlServer);
+            fnControlServer = null;
+            closeFnServer(fnDataServer);
+            fnDataServer = null;
+            closeFnServer(fnStateServer);
+            fnStateServer = null;
+          }
+        });
+  }
+
+  private void setUpFnApiServer() {
+    controlClientPool = MapControlClientPool.create();
```

Comment:
DO we need to close this pool as well as the executorService during shutdown?

Reply:
the pool doesn't have a way to close. So setting the variable to null is the best we could do. Do need to shutdown the executorService though. Good point